Dynamic Forecast.Solar location option

bencorrado opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
No breaking changes, existing configurations will not evaluate as TRUE for the dynamic_location flag and use the fixed location that is currently in use in this component.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
There are a number of users that have solar mounted on their mobile camper/RV/caravan/car that is running home assistant.  Currently the only way to get Forecast.Solar data for your mobile installation is to remove and readd the component.  This change adds a checkbox at the setup of Forecast.Solar that asks the user if they would like dynamic updates.  When this box is checked the lat/long used to query Forecast.Solar is updated from `hass.config.latitude` and `hass.config.longitude` each time.  This allows mobile users that update their Home Assistant config with current location data to get the relevant data.  When this box is not checked, the existing `entry.data[CONF_LATITUDE]` and `entry.data[CONF_LONGITUDE]` are used.

This addresses issue #85079 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes issue: fixes #85079 
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

bencorrado committed:
Updated Forecast.Solar to support dynamic location updates with checkbox.

bencorrado committed:
Edited to only be in inital change.

bencorrado committed:
Merge branch 'dev' into dynamic_forecastsolar_location

frenck reviewed (changes requested):
I think it would make more sense to track a zone in that case?

That said, what would a dynamic location offer? The declination and azimuth will change as well 🤷

frenck reviewed (changes requested):
I think it would make more sense to track a zone in that case?

That said, what would a dynamic location offer? The declination and azimuth will change as well 🤷

bencorrado commented:
> I think it would make more sense to track a zone in that case?

I did look at zones.  [It seems quite challenging to update a zones location dynamically.](https://community.home-assistant.io/t/change-zone-depending-on-gps-location-rv-is-a-moving-zone/435611)  Maybe a device_tracker or person if you don't think we should set it to the HA config location.  Most RVers seem to update their location using the config for their RV from the forums.

> That said, what would a dynamic location offer? The declination and azimuth will change as well 🤷

Most RV/camper/caravans mount the panels with a declination and azimuth of 0 degrees as they are flat on the roof.  I know a few RVers who use Home Assistant and all of them have fixed panels flat on the roof.  With the dynamic location we can plan when to run alternate power sources or automate my loads.  Currently we have to remove and re-add the component to get updated information for each campsite, which means it does not happen.

frenck commented:
The home zone can just be changed dynamically. We have a service for that out of the box.

Alternatively, "dynamic" may just sounds too complicated, and just having an option use "My Home Location" would do.

> Most RV/camper/caravans mount the panels with a declination and azimuth of 0 degrees as they are flat on the roof. I know a few RVers who use Home Assistant; all have fixed panels flat on the roof. 

Right, I get where you are going, but IMHO, that is making too many assumptions to be viable.

bencorrado committed:
Updated label for dynamic button and ruff errors.

bencorrado committed:
Merge branch 'dynamic_forecastsolar_location' of github.com:bencorrado/core into dynamic_forecastsolar_location

bencorrado commented:
> Alternatively, "dynamic" may just sounds too complicated, and just having an option use "My Home Location" would do.

I updated the PR to include this suggestion.

frenck commented:
Right, but that doesn't solve any of the azimuth and declination problems; without that, IMHO, this PR isn't viable.

Additionally, the config flow would need to be changed into a multiple-step one. Asking for coordinates and the dynamic checkbox in the same form isn't really user-friendly. It should be a choice resulting in either.

bencorrado commented:
> Right, but that doesn't solve any of the azimuth and declination problems; without that, IMHO, this PR isn't viable.

I respectfully disagree on this for a few reasons.
1. Currently one can more easily configure the azimuth and declination using the Configure button for the component, without having to delete and re-add the whole component.  This is a significantly easier user experience than trying to change the location data.
2. For a fixed user, even if they do click the box to allow it to use their Home Location will not effect the azimuth or declination as their home is not moving, leading to no negative impact to fixed users.  
3. A significant number of the mobile solar users are a fixed 0 degree declination, which makes the azimuth irrelevant.

I suppose we could allow for a user to set a sensor to feed the azimuth and elevation data.  This would likely need to be in an advanced section of a config flow.  Being able to set this might be of interest to fixed location users too, as they are more likely to have linear actuators on their panels to change the angles.

> Additionally, the config flow would need to be changed into a multiple-step one. Asking for coordinates and the dynamic checkbox in the same form isn't really user-friendly. It should be a choice resulting in either.

This is fixable and I would be happy to figure out how to implement it.

bencorrado committed:
Multiple-step config flow & az/dec sensor support

bencorrado committed:
Merge branch 'dev' into dynamic_forecastsolar_location

bencorrado commented:
So I have modified this to create a version of this that uses multiple-step config flows to allow users to select from: Home location, a Zone (as requested by @frenck), and manual input of latitude/longitude.

To address the blocking azimuth and declination problems brought up by @frenck, I added to the multiple-step config flow the ability to select a sensor for each if a sensor with a measurement type of `degrees` exists.  For installations without this type of sensor it will only allow manual entry of azimuth and declination as it is currently done, minimizing confusion for users without sensors.  

This will allow mobile users the ability to get dynamic updates where ever they are.  It will also allow people with seasonal arrays where they change the tilt to input this with an accelerometer sensor or other tilt sensor. 

The sensors are error checked and validated before use and warnings are logged.  The processing of the request is delayed until sensors are valid using `raise UpdateFailed` to handle auto retries.  This also takes care of race conditions at startup where forecast_solar may be up before a sensor is available.

I have done testing with this using mock sensors and an accelerometer and magnetometer on my camper and it is working pretty well.  

I need to add a significant number of tests now.  I wanted to check and make sure this logic and implementation approach is acceptable before building out all of the tests.

jrigling commented:
Just adding a comment that I have solar on my RV/camper and would like this dynamic location feature. My panels are mounted permanently with 0 elevation and therefore irrelevant azimuth, but I appreciate the author's additional ability to use sensors.

bencorrado commented:
@frenck I know you are a busy man, but any thoughts on this?  I will clean everything up and make tests if this new logic works for you.  Thanks!

bencorrado commented:
@frenck I know you are a busy man, but any thoughts on this?  I will clean everything up and make tests if this new logic works for you.  Thanks!

klaasnicolaas commented:
Note that as long as your pull request is still marked as draft, no core members will look at it. So once you have processed all existing feedback, it's best to mark your pull request as "ready for review" 😉
